### PR TITLE
feat(source): throttling source based on storage stats

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Implement #7997. There will be two types of source throttlers:
- StateStore throttler, which is a new one based on hummock storage stats.
- MaxWaitBarrier throttler, which is a wrapper of original barrier latency based one.

As before
- we only try to pause source when receiving a stream chunk.
- we only try to resume source when receiving a barrier.

Note that `barrier_interval_ms` will be a mutable system param (not implemented yet). So MaxWaitBarrier throttler should be updated as well at that time.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by soundOfDestiny — APPROVED** [review]
LGTM

**Review by xx01cyx — COMMENTED** [review]
LGTM

**Comment by zwang28** [comment]
This PR adds a new sys param `throttle_l0_sub_level_number`. Being similar to `barrier_interval_ms` as a mutable one, Its change must be applied to compute node too.
Please let me know when we're going to implement mutable sys params, and we shall fix `throttle_l0_sub_level_number` then.

**Comment by xx01cyx** [comment]
IIUC, the semantics here indicates that we're not accepting any data into the stream. However, apart from the source executor, the DML executor can also yield data chunks into the stream. `pause_source` simply pauses the **source**, but not DML from the user. This should be handled in case a batch-intensive use case, but I think pausing the source is enough for now. 🤔

**Comment by xx01cyx** [comment]
> IIUC, the semantics here indicates that we're not accepting any data into the stream. However, apart from the source executor, the DML executor can also yield data chunks into the stream. `pause_source` simply pauses the **source**, but not DML from the user. This should be handled in case a batch-intensive use case, but I think pausing the source is enough for now. 🤔

BTW we should not use `Barrier::Pause` to handle this because it is designed only for scaling now. cc. @BugenZhao

**Review by xx01cyx — APPROVED** [review]

File: src/stream/src/executor/source/fs_source_executor.rs [hunk 1]
```diff
@@ -330,26 +330,27 @@ impl<S: StateStore> FsSourceExecutor<S> {
...
         let mut metric_row_per_barrier: u64 = 0;
         while let Some(msg) = stream.next().await {
             match msg? {
                 // This branch will be preferred.
                 Either::Left(barrier) => {
-                    last_barrier_time = Instant::now();
-                    if self_paused {
+                    for throttler in &mut self.throttlers {
+                        throttler.on_barrier();
+                    }
```
Comment (tabVersion) [inline src/stream/src/executor/source/fs_source_executor.rs:348]: why do we need multiple throttler in source exec

File: src/stream/src/executor/source/throttler.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+}
+
+impl SourceThrottlerImpl {
+    pub fn should_pause(&self) -> bool {
+        match self {
+            SourceThrottlerImpl::MaxWaitBarrier(inner) => inner.should_pause(),
+            SourceThrottlerImpl::StateStore(inner) => {
+                if let Some(hummock) = inner.as_hummock() {
+                    return hummock.need_write_throttling();
```
Comment (tabVersion) [inline src/stream/src/executor/source/throttler.rs:54]: any hint on when hummock needs throttle?

**Review by tabVersion — COMMENTED** [review]
basically LGTM. 
please add some log about why to throttle when throttler pauses a stream and resumes it

File: src/stream/src/executor/source/throttler.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/stream/src/executor/source/throttler.rs:54]: > any hint on when hummock needs throttle?

when `ingest_batch` finds that there are too many L0 SSTs

File: src/storage/src/hummock/mod.rs [hunk 3]
```diff
@@ -271,6 +274,18 @@ impl HummockStorage {
     pub fn get_pinned_version(&self) -> PinnedVersion {
         self.pinned_version.load().deref().deref().clone()
     }
+
+    pub fn need_write_throttling(&self) -> bool {
```
Comment (Li0k) [inline src/storage/src/hummock/mod.rs:278]: Do we need to distinguish between different groups of l0 ? , I think their write paths are independent

**Review by hzxa21 — COMMENTED** [review]
Neat! LGTM!

**Review by hzxa21 — APPROVED** [review]

File: src/storage/src/hummock/mod.rs [hunk 4]
```diff
@@ -271,6 +274,18 @@ impl HummockStorage {
...
+        self.pinned_version
+            .load()
+            .version_ref()
+            .levels
+            .values()
+            .any(|levels| {
+                levels.l0.as_ref().unwrap().sub_levels.len()
+                    > self.options.throttle_l0_sub_level_number as usize
```
Comment (Li0k) [inline src/storage/src/hummock/mod.rs:286]: Shall we explain here the problem of deadlock when throttle_l0_sub_level_number < level0_tier_compact_file_number?

**Review by tabVersion — APPROVED** [review]

File: src/stream/src/executor/source/fs_source_executor.rs [hunk 5]
```diff
@@ -386,11 +393,7 @@ impl<S: StateStore> FsSourceExecutor<S> {
                     chunk,
                     split_offset_mapping,
                 }) => {
-                    if last_barrier_time.elapsed().as_millis() > max_wait_barrier_time_ms {
-                        // Exceeds the max wait barrier time, the source will be paused. Currently
-                        // we can guarantee the source is not paused since it received stream
-                        // chunks.
-                        self_paused = true;
+                    if !stream.paused() && self.throttlers.iter().any(|t| t.should_pause()) {
                         stream.pause_source();
```
Comment (fuyufjh) [inline src/stream/src/executor/source/fs_source_executor.rs:397]: Please add `warn!` logging to inform us/users that streaming is paused due to accumulated L0 SST files.

**Comment by xx01cyx** [comment]
> IIUC, the semantics here indicates that we're not accepting any data into the stream. However, apart from the source executor, the DML executor can also yield data chunks into the stream. `pause_source` simply pauses the **source**, but not DML from the user. This should be handled in case a batch-intensive use case, but I think pausing the source is enough for now. 🤔

The DML executor is able to pause now via #8110. It pauses and resumes in a similar way to the source executor, so the new throttle can be implemented on the DML executor without much pain.

**Comment by xx01cyx** [comment]
> > The DML executor is able to pause now via #8110.
> 
> Cool. BTW why there is no barrier latency based throttler in the DML executor, just like source executor? @xx01cyx

We need this throttle on the source executor to control the data stream from external connectors, based on which we could gain control over the whole graph. BTW, as another form of data source, maybe the DML executor requires a similar throttling mechanism on user data. cc. @BugenZhao @tabVersion

**Comment by zwang28** [comment]
We have decided to stall locally (ingest_batch), instead of pausing source.

**Pull request closed without merging**